lib/aws/signature.ex

```diff
@@ -14,229 +10,19 @@ defmodule AWS.Signature do
-      ],
-      "\n"
-    )
-  end
-
-  # Generate a credential scope from a short date in `YYMMDD` format, a region
-  # identifier and a service identifier.
-  defp credential_scope(short_date, region, service) do
-    Enum.join([short_date, region, service, "aws4_request"], "/")
-  end
-
-  # Convert a list of headers to canonicals signed header format.  Leading and
-  # trailing whitespace around names is stripped, header names are lowercased,
-  # and header names are semicolon-joined in alphabetical order.
-  @spec signed_headers([{binary(), binary()}]) :: binary()
-  defp signed_headers(headers) do
-    headers
-    |> Enum.map(fn {name, _value} -> name |> String.downcase() |> String.trim() end)
-    |> Enum.sort()
-    |> Enum.join(";")
-  end
-
-  # Generate a signing key from a secret access key, a short date in `YYMMDD`
-  # format, a region identifier and a service identifier.
-  defp signing_key(%Client{} = client, short_date) do
-    ("AWS4" <> client.secret_access_key)
-    |> AWS.Util.hmac_sha256(short_date)
-    |> AWS.Util.hmac_sha256(client.region || @default_region_for_global_services)
-    |> AWS.Util.hmac_sha256(client.service)
-    |> AWS.Util.hmac_sha256("aws4_request")
-  end
-
-  # Strip the query string from the URL, if one if present, and return the URL
-  # and the normalized query string as separate values.
-  defp split_url(url) do
-    url = URI.parse(url)
-
-    {uri_encode(url.path), normalize_query(url.query)}
-  end
-
-  # Copied from https://github.com/ex-aws/ex_aws/blob/623478ed321ffc6c07fdd7236a2f0e03f1cbd517/lib/ex_aws/request/url.ex#L108
-  def uri_encode(url), do: URI.encode(url, &valid_path_char?/1)
-
-  # Space character
-  defp valid_path_char?(?\ ), do: false
-  defp valid_path_char?(?/), do: true
-
-  defp valid_path_char?(c) do
-    URI.char_unescaped?(c) && !URI.char_reserved?(c)
-  end
-
-  # Sort query params by name first, then by value (if present). Append "=" to
-  # params with missing value.
-  # Example: "foo=bar&baz" becomes "baz=&foo=bar"
-  defp normalize_query(nil), do: ""
-
-  defp normalize_query(""), do: ""
-
-  defp normalize_query(query) do
-    query
-    |> String.split("&")
-    |> Enum.map(&String.split(&1, "="))
-    |> Enum.sort(fn [a, _], [b, _] -> a <= b end)
-    |> Enum.map_join("&", fn
-      [key, value] -> key <> "=" <> value
-      [key] -> key <> "="
-    end)
-  end
-
-  # Generate the text to sign from a long date in `YYMMDDTHHMMSSZ` format, a
-  # credential scope and a hashed canonical request.
-  defp string_to_sign(long_date, credential_scope, hashed_canonical_request) do
-    Enum.join(
-      ["AWS4-HMAC-SHA256", long_date, credential_scope, hashed_canonical_request],
-      "\n"
+    :aws_signature.sign_v4(
+      client.access_key_id,
+      client.secret_access_key,
+      region,
+      client.service,
+      {{now.year, now.month, now.day}, {now.hour, now.minute, now.second}},
+      to_string(method),
+      url,
+      headers,
+      body
     )
```

Comment:
We should pass `uri_encode_path: false` if the service is S3. But this should go together with the path escaping fix I mentioned offline.

Example: given S3 key `"my$key"`, it should be encode into URL like `"/my%24key"` before passing it to `sign_v4`. Then for S3 we disable the additional encoding and it stays as `"/my%24key"` when normalizing, but for any other services it's further encoded as `"/my%2524key"`.

Reply:
I think I got it right. Please check again.

Reply:
Looks great! I think the change should also be reflected in the codegen package.

Reply:
https://github.com/aws-beam/aws-codegen/commit/31b38916de305fa0c6257ba123f21f1894793b2c

Reply:
Awesome!